### ISSUE TITLE
Missing issuer option on NPM, outdated package

### ISSUE THREAD

**Issue by Cheprer** [speakeasyjs/speakeasy#86]
I've installed speakeasy module from npm but the index.js is older then in master and referenced in readme.md file. There is missing issuer option to be included in the otp path url. Can you update it please?

Or is there a version to somewhere 

Thank you,
Jan

**Comment by railsstudent** [comment]
The git repo includes issuer option in generateSecret function.

Can you specify "speakeasy": "git://github.com/speakeasyjs/speakeasy.git#e63b936d562fa16c5970d0821a221cf1a59f559b" in package.json to pull the latest source code and pass issuer in generateSecret function?

The wiki https://github.com/speakeasyjs/speakeasy/wiki/General-Usage-for-Time-Based-Token
includes an example to pass issuer value to generateSecret.

**Comment by Cheprer** [comment]
hey @railsstudent thanks for the answer I've tried the code version you've sent and it works. When we can expect to be available through npm?

**Comment by railsstudent** [comment]
What do you mean by wrong? 
The verify function does not return correct value when code is passed in?

**Comment by Cheprer** [comment]
Verify function returns correct value but scanned QR code with Google Authenticator generates the wrong value.

**Comment by jakelee8** [comment]
The secret values are different. Regression?

**Comment by BrandonCopley** [comment]
Here's the problem: when you run the code "generateSecret()" and pass in an "issuer" as part of the options, the issuer gets dropped due to the following lines:

Line 474 in index.js

    // add in the Google Authenticator-compatible otpauth URL
      if (otpauth_url) {
        SecretKey.otpauth_url = exports.otpauthURL({
          secret: SecretKey.ascii,
          label: name
        });
      }

As you can see, even if you pass in "issuer" the issuer won't be passed to the otpauthurl api that generates the google authenticator compatible otpauth URL.

**Comment by BrandonCopley** [comment]
Ahh, this fix appears to already exist in the current branch - what can we do to get this pushed up to NPM?

**Comment by jakelee8** [comment]
@Cheprer @BrandonCopley Have time to try https://github.com/speakeasyjs/libotp ?

**Comment by Cheprer** [comment]
@jakelee8 as @BrandonCopley mentioned the problem is that the issuer is not passed to otpauthURL constructor so when I've modified code to something like this then it works (of course it needs some check and defaults on the beginning):

``` javascript
if (otpauth_url) {
    SecretKey.otpauth_url = exports.otpauthURL({
      secret: SecretKey.ascii,
      label: name,
      issuer: options.issuer
    });
  }
```

works for both https://github.com/speakeasyjs/libotp and https://github.com/speakeasyjs/speakeasy

**Comment by Cheprer** [comment]
@railsstudent @jakelee8 hey guys I can see this code correctly fixed on the repository on master branch but the NPM package doesn't have it. Can you please update NPM package please and close this issue? Much appreciated, Cheprer

**Comment by jakelee8** [comment]
@markbao

**Comment by max-arias** [comment]
+1

**Comment by guillaumev** [comment]
+1

**Comment by MrXyfir** [comment]
+1

**Comment by BarryCarlyon** [comment]
+1 it's not on NPM yet

**Comment by BarryCarlyon** [comment]
Also:

if you take the response to generateSecret and then make your own URL you have to pre-encode the label your self as the library isn't. To answer @Cheprer 

Which as he reports results in:

![image](https://user-images.githubusercontent.com/20999/30773444-97228dee-a068-11e7-8993-8859dffc433d.png)

The correct function call is:

```
var secret = speakeasy.generateSecret({
    length: 20,
    name: 'Barry Carlyon',
    issuer: 'Some Issuer'
});

secret.otpauth_url = speakeasy.otpauthURL({
    secret: secret.ascii,
    label: encodeURIComponent('Barry Carlyon'),
    issuer: 'Some Issuer'
});
```

The fault when scanning the QR code only failing as the URL contains unencoded spaces

For example:

> node create_user.js
> otpauth://totp/Barry%20Carlyon?secret=SOMESCRET
> otpauth://totp/Barry Carlyon?secret=SOMESCRET&issuer=Some%20Issuer

**Comment by jakelee8** [comment]
@markbao controls the NPM package

**Comment by Cheprer** [comment]
@BarryCarlyon thanks for wrapping it up. I hope guys will make it stable and put on NPM soon.

**Comment by LukeXF** [comment]
> The fault when scanning the QR code only failing as the URL contains unencoded spaces

I'm having the same issue. A string with a space in breaks the url from otpauthURL.

**Comment by BarryCarlyon** [comment]
@LukeXF I answered the problem in my solution code:

```
secret.otpauth_url = speakeasy.otpauthURL({
    secret: secret.ascii,
    label: encodeURIComponent('Barry Carlyon'),
    issuer: 'Some Issuer'
});
```

**Comment by LukeXF** [comment]
It still adds in %20 with `encodeURIComponent()`

**Comment by railsstudent** [comment]
@LukeXF  That what encodeURIComponent() does, it  replaces space with %20

**Comment by LukeXF** [comment]
@railsstudent Yes, but the %20 is showing up on Google Authenticator. `Project%20Name` for example instead of `Project Name`

**Comment by jakelee8** [comment]
Sounds like double escaping

**Comment by markbao** [comment]
@LukeXF Are you using the npm 2.0.0 version of the package, or the latest master from git?

In the npm version, label is not URL-encoded: https://github.com/speakeasyjs/speakeasy/blob/4e0aa91dee166306e07ea11e371d5a5df95f4391/index.js#L644

In the latest master from git, it is: https://github.com/speakeasyjs/speakeasy/blob/master/index.js#L704

In the latest master from git, issuer *should* be URL-encoded, since we pass in an object to `url.format` and that runs `querystring`'s `stringify`.

**Comment by LukeXF** [comment]
@markbao I'm using NPM 2.0.0. But even with using `encodeURIComponent()` myself, the `%20` still outputs as `%20` and not a space. Any suggestions?

**Comment by LukeXF** [comment]
Problem was with base32 encoding. Because I'm using a base32 secret, I needed to specify an encoding so that it didn't load ASCII be default.

See https://github.com/speakeasyjs/speakeasy/issues/95#issuecomment-334955321 for reference.

```
var url = speakeasy.otpauthURL({
	secret: secret.base32,
	label: req.user.local.email,
	issuer: 'The Spaces Work',
	encoding: "base32"
});
```

**Comment by Cheprer** [comment]
I've used @LukeXF 's way and it works. Big thanks but even though the function generateSecret doesn't work properly still because it generates ASCII secret and then from it it generates the base32. Maybe there is a glitch. If I'll have time. I'll try to make a pull request with fix but not sure when. So I would keep this issue still open.

**Comment by joaonice** [comment]
@markbao any plans on releasing a new version where the `issuer` is passed to the `otpauthURL` function on `generateSecret`?

**Comment by mikgross** [comment]
Hi all, old issue but the issuer still misses in NPM on the generateSecret() method. Any plans to implement it?

**Comment by nrail-joerg-walter** [comment]
This is still in the NPM package. Please update!

**Comment by jlous** [comment]
No releases i 4 years, no progress on v.3 i 3 years.  Is project dead?

**Comment by carvalholeo** [comment]
I think yes, @jlous. @mlogan did a fork, but I think that project is dead too. I tried to contact @markbao on Twitter, but he didn't answer me, so I guess that could be great if someone fork again and implement these features/bug fixes.

**Comment by jlous** [comment]
Alternatively, https://www.npmjs.com/package/otplib seems to be alive and well.
Will consider migrating my code.
